In [43]:
#importing data handling libraries
import pandas as pd
import numpy as np

#Importing label encoder
from sklearn.preprocessing import LabelEncoder

#Importing train test split
from sklearn.model_selection import train_test_split

#Importing models
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

#Importing metrics for evaluation
from sklearn.metrics import accuracy_score

In [29]:
#Reading and displaying data
df = pd.read_excel('E:/Businesstoys/ktm.xlsx')
df.head()

ID  Age  Gender     Occupation Phone Type Current Bike Relationship  \
0   1   53   Male    Professional    Average   180 to 220  Complicated   
1   2   27  Female  Self Employed    Low End      No Bike       Single   
2   3   39  Female     Unemployed    Average   180 to 220      Married   
3   4   20  Female     Unemployed   High End      No Bike      Married   
4   5   29   Male         Student    Average   180 to 220  Complicated   

        Response  
0  Not purchased  
1      Purchased  
2  Not purchased  
3  Not purchased  
4      Purchased

In [30]:
df = df.drop(columns=['ID'], axis=1) #dropping Id column

In [31]:
df.head()

Age  Gender     Occupation Phone Type Current Bike Relationship  \
0   53   Male    Professional    Average   180 to 220  Complicated   
1   27  Female  Self Employed    Low End      No Bike       Single   
2   39  Female     Unemployed    Average   180 to 220      Married   
3   20  Female     Unemployed   High End      No Bike      Married   
4   29   Male         Student    Average   180 to 220  Complicated   

        Response  
0  Not purchased  
1      Purchased  
2  Not purchased  
3  Not purchased  
4      Purchased

## Converting categorical to continous using Labelencoder

In [36]:
df.dtypes

Age              int64
Gender          object
Occupation      object
Phone Type      object
Current Bike    object
Relationship    object
Response        object
dtype: object

In [5]:
le = LabelEncoder()

In [6]:
cols = df.columns
cols

Index(['Age', 'Gender', 'Occupation', 'Phone Type', 'Current Bike',
       'Relationship', 'Response'],
      dtype='object')

In [27]:
df.dtypes

Age             int64
Gender          int32
Occupation      int32
Phone Type      int32
Current Bike    int32
Relationship    int32
Response        int32
dtype: object

In [40]:
#Udf for label encoding categorical into continous.
def label_enc():
    for col in df.columns:
        if df[col].dtype == 'O':
            df[col]= le.fit_transform(df[col])

In [41]:
label_enc()

In [11]:
# df['Occupation']= le.fit_transform(df['Occupation'])
# df['Phone Type']= le.fit_transform(df['Phone Type'])
# df['Current Bike']= le.fit_transform(df['Current Bike'])
# df['Relationship']= le.fit_transform(df['Relationship'])
# df['Response']= le.fit_transform(df['Response'])
# df['Gender']= le.fit_transform(df['Gender'])

In [12]:
df.head()

Age  Gender  Occupation  Phone Type  Current Bike  Relationship  Response
0   53       1           0           0             1             1         0
1   27       0           1           2             4             3         1
2   39       0           3           0             1             2         0
3   20       0           3           1             4             2         0
4   29       1           2           0             1             1         1

In [13]:
#Features
x = df.drop('Response', axis=1)
x.head()

Age  Gender  Occupation  Phone Type  Current Bike  Relationship
0   53       1           0           0             1             1
1   27       0           1           2             4             3
2   39       0           3           0             1             2
3   20       0           3           1             4             2
4   29       1           2           0             1             1

In [14]:
#target
y = df[['Response']]
y.head()

Response
0         0
1         1
2         0
3         0
4         1

In [23]:
#creating udf that will train, fit and evaluate the model
def my_func(x,y, model):
    xtrain, xtest, ytrain, ytest = train_test_split(x,y, test_size=0.3)
    model = model()
    model.fit(x, y)
    model.fit(xtrain, ytrain)
    pred = model.predict(xtest)
    acc_te = accuracy_score(ytest, pred)
    pred1 = model.predict(xtrain)
    acc_tr = accuracy_score(ytrain, pred1)
    print(f'Testing accuracy is {acc_te}')
    print(f'Training accuracy is {acc_tr}')

In [44]:
my_func(x, y, model=GradientBoostingClassifier)

C:\Users\kaush\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
C:\Users\kaush\anaconda3\lib\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)


Testing accuracy is 0.7807017543859649
Training accuracy is 0.8559322033898306


For SVC
Testing accuracy is 0.7960526315789473
Training accuracy is 0.7608286252354048

For Random forest
Testing accuracy is 0.7785087719298246
Training accuracy is 0.995291902071563

For Adaboost
Testing accuracy is 0.793859649122807
Training accuracy is 0.7909604519774012

For DecisionTreeClassifier
Testing accuracy is 0.7653508771929824
Training accuracy is 0.9962335216572504


Testing accuracy is 0.7807017543859649
Training accuracy is 0.8559322033898306